src/locale/keymap-util.c

```diff
@@ -780,3 +781,137 @@ int x11_convert_to_vconsole(Context *c) {
 
         return modified;
 }
+
+bool locale_gen_check_available(void) {
+        struct stat st;
+        if (stat("/usr/sbin/locale-gen", &st) < 0)
+                return false;
+        if (stat("/etc/locale.gen", &st) < 0)
+                return false;
+        return true;
+}
+
+int locale_gen_enable_locale(char *locale[_VARIABLE_LC_MAX]) {
+        _cleanup_fclose_ FILE *fr = NULL;
+        _cleanup_fclose_ FILE *fw = NULL;
+        _cleanup_free_ char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false;
+        bool first_line;
+        struct stat st;
+        int r;
+        const char *LOCALE_GEN_CONF_FNAME = "/etc/locale.gen";
+
+        if (isempty(locale[VARIABLE_LANG]))
+                return -EINVAL;
+
+        if (stat(LOCALE_GEN_CONF_FNAME, &st) < 0)
+                return -errno;
+
+        fr = fopen(LOCALE_GEN_CONF_FNAME, "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary(LOCALE_GEN_CONF_FNAME, &fw, &temp_path);
+        if (r < 0)
+                return r;
+        (void) fchmod(fileno(fw), 0644);
+
+        if (!strstr(locale[VARIABLE_LANG], ".") || endswith_no_case(locale[VARIABLE_LANG], "UTF-8")) {
+                if (asprintf(&locale_entry, "%s UTF-8", locale[VARIABLE_LANG]) < 0)
+                        return -ENOMEM;
+        } else {
+                /* We do not process non-UTF-8 locale */
+                return -EINVAL;
+        }
```

Comment:
why open the temporary file and evertyhing if we are returning here anyway? please do this test early on so that no unnecessray work is done.

also, we generally don't place {} around single-line if blocks.